### PR TITLE
HL Python SDK: Fix transaction cleanup

### LINKED ISSUE #7214: [Bug]: Python HL SDK -  transaction branches not deleted if no modifications made
### What happened?

What actually happened, including error codes if applies.
        
Steps to Reproduce:
1. create a transaction
2. use it to read data, but don't write anything
3. let the transaction end.

### Expected behavior

The underlying transaction branch should be removed since there was no error during the processing of the transaction. 

Example:

```python
with branch.transact(commit_message='remove expired data') as tx:
    expired_objects = json.load(tx.object('manifest.json').reader())
    if expired_objects.get('should_delete'):
        tx.delete_objects(expired_objects.get('paths'))
```

This transaction has a condition inside it that might result in no modifications. As expected, there's no error and no exception is thrown, however - the transaction branch is not deleted once the transaction is complete.

### lakeFS version

1.4

### How lakeFS is installed

locally, cloud

### Affected clients

Python: lakefs==0.2.0

### PR DESCRIPTION
**Pull request by N-o-Z**
Closes #7214

### Background

If no changes occur on tx branch - we skip the exit flow and as part of it we do not delete the branch

### Bug Fix

Cleanup branch even if no changes were made
      
### Testing Details

Modified integration test

### CODE REVIEW

File: clients/python-wrapper/lakefs/branch.py [hunk 1]
```diff
@@ -356,11 +356,10 @@ def __enter__(self):
...
         try:
-            if len(list(self._tx.uncommitted(amount=1))) == 0:
-                return  # Do nothing if tx is empty
-
-            self._tx_branch.commit(message=self._tx.commit_message, metadata=self._tx.commit_metadata)
-            self._tx_branch.merge_into(self._source_branch, message=f"Merge transaction {self._tx.id} to branch")
-            self._tx_branch.delete()
+            if len(list(self._tx.uncommitted(amount=1))) > 0:
+                self._tx_branch.commit(message=self._tx.commit_message, metadata=self._tx.commit_metadata)
+                self._tx_branch.merge_into(self._source_branch, message=f"Merge transaction {self._tx.id} to branch")
```
Comment (guy-har) [inline clients/python-wrapper/lakefs/branch.py:361]: This does a bit more than just clean up a branch, it ignores an error, is this expected? I would prefer if we have a special case for errors of type `no-changes` and remove the branch in that case.

**Review by guy-har — CHANGES_REQUESTED** [review]
Change looks good, I'm requesting changes because I'm not sure it meets the requirement, TBH I'm not sure what the requirement is.
By this implementation, an empty transaction will not throw an error, but will not add any commit either. I would expect one of:
- Throw an error and delete the branch
- Create an empty commit with the requested message and metadata and delete the branch

File: clients/python-wrapper/lakefs/branch.py (hunk 1 above)
Comment (N-o-Z) [reply clients/python-wrapper/lakefs/branch.py:364]: Though the code was logically correct, I moved it back into the try.
Can you elaborate on what errors are we ignoring? I don't think we need to treat no-changes as an error.
What is missing is dealing with exceptions raised while inside the context - which I added now. Thanks!

File: clients/python-wrapper/CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,10 @@
 
 ## Unreleased
 
+:bug: Bugs fixed:
+
+- Fix transaction cleanup (#7227)
```
Comment (ozkatz) [inline clients/python-wrapper/CHANGELOG.md:7]:
```suggestion
- Cleanup transaction branches that weren't modified (#7227)
```

File: clients/python-wrapper/lakefs/branch.py [hunk 3]
```diff
@@ -354,13 +354,16 @@ def __enter__(self):
...
 
-    def __exit__(self, exc_type, exc_val, exc_tb):
+    def __exit__(self, typ, value, traceback) -> bool:
+        if typ is not None:  # Do nothing in case exception occurred
+            return False  # Raise the underlying exception
+
         try:
-            if len(list(self._tx.uncommitted(amount=1))) == 0:
-                return  # Do nothing if tx is empty
+            if len(list(self._tx.uncommitted(amount=1))) > 0:
```
Comment (ozkatz) [inline clients/python-wrapper/lakefs/branch.py:362]:
```suggestion
            if list(self._tx.uncommitted(amount=1)):
```

**Review by ozkatz — APPROVED** [review]
Looks so good I had to nitpick :)

File: clients/python-wrapper/lakefs/branch.py [hunk 4]
```diff
@@ -354,13 +354,19 @@ def __enter__(self):
...
 
-    def __exit__(self, exc_type, exc_val, exc_tb):
-        try:
-            if len(list(self._tx.uncommitted(amount=1))) == 0:
-                return  # Do nothing if tx is empty
+    def __exit__(self, typ, value, traceback) -> bool:
+        if typ is not None:  # Do nothing in case exception occurred
+            return False  # Raise the underlying exception
+
+        if len(list(self._tx.uncommitted(amount=1))) == 0:  # Delete branch and raise exception
```
Comment (guy-har) [inline clients/python-wrapper/lakefs/branch.py:361]: In my previous comment, I meant without the need to list uncommitted, just react to the type of error returned when trying to commit and merge. Theoretically, someone can delete the changes from the branch between the list and commit, and we will still return no changes and not remove the branch

**Review by guy-har — APPROVED** [review]
See my comment

**Review by guy-har — APPROVED** [review]
LGTM

**Pull request merged**